### PR TITLE
Explore Agent's rewrite options

### PR DESCRIPTION
**Pull request by masci**
### What does this PR do?

Design proposal for a POC of a new Agent. The prototype would help us assessing the current development cycle is the best we can get and will let us plan in advance in case we want to steer the project in a different direction.

### CODE REVIEW

File: docs/proposals/rewrite.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+## Problem
+
+The real problem will be finding approval and resources for the project but let's
+not think about this for a moment.
+
+## Constraints
+
+1. Solution must be packaged with Omnibus Software
+2. Agent must run integrations in Go and Python
```
Comment (olivielpeau) [inline docs/proposals/rewrite.md]: what about Lua?

File: docs/proposals/rewrite.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+- Cost
+    - One team of 2 or 3 engineers working fulltime for 4 to 6 quarters
+
+## Other Solutions
+
+###  Use Crystal
+
+- Strengths
+    - fast as C, slick as Ruby
+    - its Ruby-like syntax will make Omnibus specialists feel at home
```
Comment (olivielpeau) [inline docs/proposals/rewrite.md]: ➕

**Review by olivielpeau — COMMENTED** [review]
Looks good!

One oversight though IMO: we should also seriously consider switching to a pure functional programming language for the Agent. I'm concerned that going with an imperative language will not create as much hype as we'd want.

File: docs/proposals/rewrite.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+a different software architecture. It was fun to do, let's do it again.
+
+## Problem
+
+The real problem will be finding approval and resources for the project but let's
+not think about this for a moment.
+
+## Constraints
+
+1. Solution must be packaged with Omnibus Software
```
Comment (hkaj) [inline docs/proposals/rewrite.md]: It's great to see this in the requirements. that's definitely the #1 requirement for a smooth developer experience 💯

File: docs/proposals/rewrite.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+- Strengths
+    - better than Java
+- Weaknesses
+    - fully interoperable with Java
+- Cost
+    - we'd pay in engineers retention, not money
+
+## Open Questions
+
+*Note any big questions that don’t yet have an answer that might be relevant.*
```
Comment (hkaj) [inline docs/proposals/rewrite.md]: I'm surprised we haven't considered C, it seems like a strong candidate for a full rewrite of the agent. Plus its hype score is very stable: https://trends.google.com/trends/explore?date=all&geo=US&q=c

File: docs/proposals/rewrite.md (hunk 2 above)
Comment (xlucas) [inline docs/proposals/rewrite.md]: What about crystal's release cycle? It changes fast and regularly ships breaking changes (https://github.com/crystal-lang/crystal/releases). Looks like a red light to me.

**Pull request closed without merging**